File: src/main/java/Naturaleza.java
public abstract class Naturaleza {
    protected static Naturaleza ventaja;
    void setVentaja(Naturaleza naturaleza)  {
        Naturaleza.ventaja = naturaleza;
    }
    Naturaleza getVentaja() {
        return Naturaleza.ventaja;
    }
    protected boolean noHayInstancia() {
        return Naturaleza.ventaja == null;
    }
}

class Robot extends Naturaleza{
    @Override
    public Naturaleza getVentaja() {
        if(this.noHayInstancia()){
            this.setVentaja(new Animal());
        }
        return super.getVentaja();
    }
}
class Animal extends Naturaleza{
    @Override
    public Naturaleza getVentaja() {
        if(this.noHayInstancia()){
            this.setVentaja(new Cosa());
        }
        return super.getVentaja();
    }
}
class Cosa extends Naturaleza{
    @Override
    public Naturaleza getVentaja() {
        if(this.noHayInstancia()){
            this.setVentaja(new Robot());
        }
        return super.getVentaja();
    }
}

File: python/Naturaleza.py
from Logger import singleton

class Naturaleza():
    def __init__(self):
        self.__ventaja = 0

    def set_ventaja(self, naturaleza) :
        self.__ventaja = naturaleza
    
    def get_ventaja(self):
        return self.__ventaja
    
    def hay_instancia(self):
        return self.__ventaja != None

@singleton    
class Robot(Naturaleza):

    def __init__(self):
        super().__init__()

    def get_ventaja(self):
        if not self.hay_instancia():
            self.set_ventaja(Animal())        
        return super().get_ventaja()

@singleton    
class Animal(Naturaleza):

    def __init__(self):
        super().__init__()
    
    def get_ventaja(self):
        if not self.hay_instancia():
            self.set_ventaja(Cosa())        
        return super().get_ventaja()

@singleton
class Cosa(Naturaleza):  

    def __init__(self):
        super().__init__()
  
    def get_ventaja(self):
        if not self.hay_instancia():
            self.set_ventaja(Robot())        
        return super().get_ventaja()